### PR TITLE
Cleanup the softmax implementation and cub functions.

### PR DESCRIPTION
**Pull request by trivialfis**
- Group cub function calls.
- Use linalg for the softmax implementation.
- Use valid label values for obj tests.

Extracted from https://github.com/dmlc/xgboost/pull/11656 .
Ref https://github.com/dmlc/xgboost/issues/9043

### CODE REVIEW

**Comment by trivialfis** [comment]
cc @rongou

**Review by rongou — APPROVED** [review]

File: src/common/algorithm.cuh [hunk 1]
```diff
@@ -327,6 +329,24 @@ void InclusiveSum(Context const *ctx, InputIteratorT d_in, OutputIteratorT d_out
 #endif
 }
 
+template <typename... Args>
+void RunLengthEncode(dh::CUDAStreamView stream, Args &&...args) {
+  std::size_t n_bytes = 0;
+  dh::safe_cuda(
+      cub::DeviceRunLengthEncode::Encode(nullptr, n_bytes, std::forward<Args>(args)..., stream));
```
Comment (trivialfis) [inline src/common/algorithm.cuh]: Remove forward

**Comment by trivialfis** [comment]
https://github.com/dmlc/xgboost/actions/runs/17274795128/job/49029841092?pr=11663
